src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -46,11 +47,16 @@ private static IEnumerable<TSource> WhereIterator<TSource>(IEnumerable<TSource>
             if (selector == null) throw Error.ArgumentNull("selector");
             Iterator<TSource> iterator = source as Iterator<TSource>;
             if (iterator != null) return iterator.Select(selector);
-            TSource[] array = source as TSource[];
-            if (array != null) return new WhereSelectArrayIterator<TSource, TResult>(array, null, selector);
-            List<TSource> list = source as List<TSource>;
-            if (list != null) return new WhereSelectListIterator<TSource, TResult>(list, null, selector);
-            return new WhereSelectEnumerableIterator<TSource, TResult>(source, null, selector);
+            IList<TSource> ilist = source as IList<TSource>;
+            if (ilist != null)
+            {
+                TSource[] array = source as TSource[];
+                if (array != null) return new SelectArrayIterator<TSource, TResult>(array, selector);
+                List<TSource> list = source as List<TSource>;
+                if (list != null) return new SelectListIterator<TSource, TResult>(list, selector);
```

Comment:
Why the specialization for ```List<T>```?  I assume this was done purely for performance, e.g. avoiding the interface calls?  If so, what kinds of gains do you see?